### DISCUSSION TITLE
Scripting hooks

### DISCUSSION THREAD

**Discussion by AdamYellen** [volllly/rotz#54] [Ideas]
For some applications it's not enough to link some files. Along the style of RPM spec files, it would be useful to run custom scripts pre- and post- install and link actions. Rotz could check the exit code of those scripts to abort further action.

For some apps I would use the pre-install to check if the app is already installed.

**Comment by volllly** [comment]
Hey, I'm not sure about that as I don't want Rotz to become a full package manager. It should mainly be used in conjunction with package managers where possible (Use an actual package manager to install the packages. E.g. `installs: apt-get install git -y`).

In the future I want to track which files are currently linked so that Rotz can remove orphaned files. Here we could also track which applications are already installed and skip installing them again (and maybe use a `--force` flag to reinstall them anyways).

Right now you can already achieve pre and post install hook functionality by using multi line scripts as the install cmd.

> Example:
> ```yaml
> installs: |
>   if(<install condition>) 
>     <install command>
>   }
>   <post command>
> ```

It is also possible that it's fine if the install command is executed again. E.g. when the application is installed via [`scoop`](https://scoop.sh/) it fails the installation and Rotz continues to the next application.

For linking hooks could you give me an example where they would be needed? I get that an inbuilt solution would be nicer but this could also be achieved by using make, [just](https://github.com/casey/just) or something similar.

**Reply by kapad** [reply]
> In the future I want to track which files are currently linked so that Rotz can remove orphaned files.

I'm starting a rewrite of a great big mess that I called dotfiles. And Rotz seemed like the perfect solution with one exception. i.e. tracking linked files and deleting the unused ones. (Is there a separate discussion on the topic? - this next para would be more relevant to it.)  

_I think this can get challenging if I am playing around with dotfiles. Say, in my mainline, I rely on the defaults for a particular program, but I have a feature branch to my dotfiles with some config for the same program that I'm testing out. The ideal workflow would be for me to wire git hooks to Rotz, and then have Rotz do the heavy lifting. So, when I change the branch to `main`, the file is removed, and then when I change the branch to `feature` the file gets created._

Lastly, I don't see myself ever using the installs functionality in Rotz. I plan to use homebrew bundle for that. 

Also, I think it's desirable to maintain separate install scripts (per OS if needed) outside of my configuration.

**Reply by volllly** [reply]
Hey, handling orphaned files is actually [already implemented](https://github.com/volllly/rotz/releases/tag/v0.9.0).

I should probably mention this in the docs ^^

**Reply by volllly** [reply]
We basically save the currently linked files and when running the link command the old and new state is compared, orphaned links are removed and everything else is linked .

**Reply by kapad** [reply]
> We basically save the currently linked files and when running the link command the old and new state is compared, orphaned links are removed and everything else is linked .

Where is this state being saved? And if I execute `rotz link` in git hooks as described in my previous comment, would switching branches also update my dotfiles (including deletion)?

**Reply by kapad** [reply]
And thanks for your prompt reply :D

**Reply by volllly** [reply]
The state file should be in these locations
|||
|---|---|
| linux | $XDG_DATA_HOME or $HOME/.local/share/rotz/state.yaml |
| windows | $HOME\AppData\Local\rotz\state.yaml |
| macos | $HOME/Library/Application Support/com.rotz/state.yaml |

And what you described should work yes 😁

**Reply by kapad** [reply]
Awesome. Thanks for all the info. :D

**Comment by AdamYellen** [comment]
I agree that this idea comes close to encroaching on package manager territory. I didn't realize that multi-line commands were allowed so that might be enough. What's the syntax?

One example of a linking script hook would be to restart an app after linking.

**Comment by AdamYellen** [comment]
@volllly How would I do something like this?
```
windows:
  installs: |
    winget install Google.Chrome --silent --accept-source-agreements --accept-package-agreements | Out-Null
    if (Test-Path -Path "C:\Users\Public\Desktop\Google Chrome.lnk") {
      Remove-Item "C:\Users\Public\Desktop\Google Chrome.lnk"
    }
```

I get the following:
```
Error: install::command::run

  x Install command for apps did not run successfully
  |-> Cloud not spawn command
  `-> program not found
```

(maybe a typo there: Cloud/Could)

**Reply by volllly** [reply]
Hey, I think you need to set the `shell_command` config key to `powershell -NoProfile -C {{ quote "" cmd }}`. (`rotz -h` shows the default path of the config file.)

I wrongly set the default to powershell core (`pwsh -NoProfile -C {{ quote "" cmd }}`) which is not installed preinstalled on windows. I'll fix this in the next release.

**Comment by AdamYellen** [comment]
That was exactly it. I tried it on a brand new install so PowerShell Core hadn't been installed yet.

Thank you!

**Reply by volllly** [reply]
Great 😁 I've just published the fix in v0.6.1.

So you should be able to remove this from your config again.
